src/Mutator/Removal/ConcatOperandRemoval.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * This code is licensed under the BSD 3-Clause License.
+ *
+ * Copyright (c) 2017, Maks Rafalko
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Mutator\Removal;
+
+use function assert;
+use Infection\Mutator\Definition;
+use Infection\Mutator\GetMutatorName;
+use Infection\Mutator\Mutator;
+use Infection\Mutator\MutatorCategory;
+use PhpParser\Node;
+
+/**
+ * @internal
+ */
+final class ConcatOperandRemoval implements Mutator
+{
+    use GetMutatorName;
+
+    public static function getDefinition(): ?Definition
+    {
+        return new Definition(
+            <<<'TxT'
+Removes an operand from a string concatenation.
+
+```php
+$x = 'foo' . 'bar';
+```
+
+Will be mutated to:
+
+```php
+$x = 'foo';
+```
+
+And:
+
+```php
+$x = 'bar';
+
+TxT
+            ,
+            MutatorCategory::SEMANTIC_REDUCTION,
+            null
+        );
+    }
+
+    public function canMutate(Node $node): bool
+    {
+        return $node instanceof Node\Expr\BinaryOp\Concat;
+    }
+
+    /**
+     * @psalm-mutation-free
+     */
+    public function mutate(Node $node): iterable
+    {
+        assert($node instanceof Node\Expr\BinaryOp\Concat);
```

Comment:
I think this should be `Assert::isInstanceOf($node, Node\Expr\BinaryOp\Concat::class);`

Reply:
`use Webmozart\Assert\Assert;`

Reply:
I'd keep the `assert()` since perf penalty in production. We're not really checking the type on other mutators as well. What do you think? (Ideally, the `Mutator` interface would be `@template`d and the type of the argument would be the template type.)

Reply:
Aye, you can use just `@param Node\Expr\BinaryOp\Concat $node`

Reply:
Just came to realize that "production" is silly when talking about a development tool. But still, I think that if we just want to help out the static analyzers (in this case PHPStan), we should use `assert()` instead of "real" code.

But I'd be happy to use the `Assert` class if you have a different opinion. Just wanted to share mine.

Reply:
I see that virtually every `::mutate()` uses `@param` to deal with the issue.

Reply:
Yes, and they're all in PHPStan's ignore list. I wanted to avoid adding it there because there's a reason PHPStan complains about it.

Reply:
Again, I think the "real" solution would be to add `@template TNode` to `Mutator` and change the argument type to `TNode` (but keep the type `Node` on `canMutate`, of course.)

Reply:
OK, how do you think if we do `make phpstan-baseline` to ignore the issue, use `@param` here, and deal with this later in a centralized fashion?

Reply:
You can ignore changes to `phpstan-tests-baseline.neon` for time being.

Reply:
I think PHPStan is overthinking here. We didn't change the input type ([that'll be a PHP error](https://3v4l.org/TqNho)), but rather we use `@param` point out exactly which subtype we know we will have here. There's no need to do any kind of assertion because we ensure we get the type we expect in `canMutate()`